### PR TITLE
Add more Foldable methods

### LINKED ISSUE #222: Foldable (HashMap k) is slower than it could be
Currently the Foldable instance for HashMap only specifies `foldr`. While that's technically enough, it gives poor Core when we use something like `Foldable.foldl'` as it gets translated into a `foldr`, `id` gets passed a few layers down &c.

We could trivially define `Foldable.foldl' = HashMap.foldl'` which is already written, available and generates much better Core. Maybe some other methods should get the same treatment.

### PR DESCRIPTION
**Pull request by treeowl**
Implement `foldl`, `foldr'`, and `foldMap` specially. Also add
functions `foldl`, `foldr'`, `foldlWithKey`, `foldrWithKey'`,
and `foldMapWithKey`.

Closes #222

### CODE REVIEW

**Comment by treeowl** [comment]
@Fuuzetsu, I'd appreciate a review.

File: Data/HashMap/Array.hs [hunk 1]
```diff
@@ -428,6 +441,41 @@ foldr f = \ z0 ary0 -> go ary0 (length ary0) 0 z0
...
+foldl f = \ z0 ary0 -> go ary0 (length ary0 - 1) z0
+  where
+    go _ary (-1) z = z
+    go ary i z
+      | (# x #) <- index# ary i
+      = f (go ary (i - 1) z) x
+{-# INLINE foldl #-}
+
+-- We go to a bit of trouble here to avoid appending an extra mempty.
+foldMap :: Monoid m => (a -> m) -> Array a -> m
```
Comment (Fuuzetsu) [inline Data/HashMap/Array.hs]: no INLINE/INLINABLE? We can't specialise.

I think this implementation is better (the Core is nicer) but I can't really seem to measure any difference. If anything, the code is a lot shorter...

```hs
foldMap f = \ary0 -> case length ary0 of
  0 -> mempty
  len ->
    let !lst = len - 1
        go i | (# x #) <- index# ary0 i, let fx = f x =
          if i == lst then fx else fx <> go (i + 1)
    in go 0
```

What do you think?

Comment (Fuuzetsu) [reply Data/HashMap/Array.hs]:
```
benchmarked treeowl
time                 17.12 ms   (16.89 ms .. 17.29 ms)
                     0.998 R²   (0.997 R² .. 0.999 R²)
mean                 17.04 ms   (16.86 ms .. 17.21 ms)
std dev              648.6 μs   (521.7 μs .. 850.1 μs)
variance introduced by outliers: 21% (moderately inflated)

benchmarked fuuzetsu
time                 17.02 ms   (16.84 ms .. 17.21 ms)
                     0.998 R²   (0.997 R² .. 0.999 R²)
mean                 16.80 ms   (16.63 ms .. 17.00 ms)
std dev              703.4 μs   (575.9 μs .. 1.035 ms)
variance introduced by outliers: 24% (moderately inflated)
```

The two are always within some small fraction of the noise in perf, which one gets a "faster" number seems random depending on the run.

**Comment by Fuuzetsu** [comment]
I think I'd love to see some benchmarks added, that way it'd be very easy to quantify the impact of this change with before/after. Do you think you can do that?

File: Data/HashMap/Array.hs (hunk 1 above)
Comment (treeowl) [reply Data/HashMap/Array.hs]: I'm a bit concerned that lifting out `ary0` and `len` could reduce performance by forcing `go` to be allocated as a closure. How do benchmarks compare for very small and very large maps? I did not intend to omit `INLINE`. Thanks for catching that.

Comment (treeowl) [reply Data/HashMap/Array.hs]: Summing like that is a rather bad benchmark, because you'll end up with big thunks.

**Comment by treeowl** [comment]
I hate benchmarking. :-P. But I'm willing to give it a shot.

File: Data/HashMap/Array.hs (hunk 1 above)
Comment (Fuuzetsu) [reply Data/HashMap/Array.hs]: What do you think is a good benchmark? I changed the monoid to summing with IORef over IO
```hs

firstFuuzetsu :: Array Int -> IO Int
firstFuuzetsu ar = do
    ref <- newIORef 0
    foldMapFuuzetsu (\x -> modifyIORef' ref (+ x)) ar
    readIORef ref

firstTreeOwl :: Array Int -> IO Int
firstTreeOwl ar = do
    ref <- newIORef 0
    foldMapTreeOwl (\x -> modifyIORef' ref (+ x)) ar
    readIORef ref

main :: IO ()
main = defaultMain
    [ env (mkAr len) $ \ar -> bgroup (show len)
        [ bench "treeowl" $ nfIO (firstTreeOwl ar)
        , bench "fuuzetsu" $ nfIO (firstFuuzetsu ar)
        ]
    ]
    where
        len = 100000
```

```
[nix-shell:/tmp]$ ghc -O2 Ar.hs && ./Ar -L10
[1 of 1] Compiling Main             ( Ar.hs, Ar.o )
Linking Ar ...
benchmarked 100/treeowl
time                 477.5 ns   (472.8 ns .. 481.7 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 478.9 ns   (476.3 ns .. 484.3 ns)
std dev              13.48 ns   (8.628 ns .. 22.69 ns)
variance introduced by outliers: 15% (moderately inflated)

benchmarked 100/fuuzetsu
time                 394.0 ns   (391.0 ns .. 397.8 ns)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 395.8 ns   (394.1 ns .. 398.1 ns)
std dev              7.718 ns   (6.444 ns .. 9.713 ns)


[nix-shell:/tmp]$ ghc -O2 Ar.hs && ./Ar -L10
[1 of 1] Compiling Main             ( Ar.hs, Ar.o )
Linking Ar ...
benchmarked 100000/treeowl
time                 1.136 ms   (1.107 ms .. 1.171 ms)
                     0.990 R²   (0.982 R² .. 0.996 R²)
mean                 1.077 ms   (1.064 ms .. 1.099 ms)
std dev              64.28 μs   (48.11 μs .. 91.49 μs)
variance introduced by outliers: 43% (moderately inflated)

benchmarked 100000/fuuzetsu
time                 1.005 ms   (948.6 μs .. 1.082 ms)
                     0.975 R²   (0.959 R² .. 0.994 R²)
mean                 1.064 ms   (1.030 ms .. 1.118 ms)
std dev              163.0 μs   (117.6 μs .. 244.0 μs)
variance introduced by outliers: 84% (severely inflated)


[nix-shell:/tmp]$ ghc -O2 Ar.hs && ./Ar -L10
[1 of 1] Compiling Main             ( Ar.hs, Ar.o )
Linking Ar ...
benchmarked 1000000/treeowl
time                 15.69 ms   (15.41 ms .. 16.04 ms)
                     0.995 R²   (0.989 R² .. 0.998 R²)
mean                 15.23 ms   (14.93 ms .. 15.50 ms)
std dev              792.1 μs   (590.5 μs .. 1.168 ms)
variance introduced by outliers: 24% (moderately inflated)

benchmarked 1000000/fuuzetsu
time                 14.50 ms   (14.12 ms .. 14.86 ms)
                     0.996 R²   (0.993 R² .. 0.998 R²)
mean                 14.29 ms   (14.00 ms .. 14.46 ms)
std dev              653.9 μs   (473.8 μs .. 1.064 ms)
variance introduced by outliers: 20% (moderately inflated)
```

Full benchmark:
https://gist.github.com/Fuuzetsu/777b6958cae9916fde32796beb26862c

Comment (Fuuzetsu) [reply Data/HashMap/Array.hs]: If I measure something that makes the inner worker tail recursive such as the below, my implementation consistently wins out. For example

```hs

newtype U = U ()
instance Semigroup U where
    {-# INLINE (<>) #-}
    _ <> y = y
instance Monoid U where
    {-# INLINE mempty #-}
    mempty = U ()
    {-# INLINE mappend #-}
    mappend = (<>)

mkAr :: Int -> IO (Array Int)
mkAr l = fromList l <$> replicateM l randomIO
    where
        fromList :: Int -> [a] -> Array a
        fromList n xs0 = run $ do
            mary <- new_ n
            go xs0 mary 0
          where
            go [] !mary !_   = return mary
            go (x:xs) mary i = do write mary i x
                                  go xs mary (i+1)

cu :: IO () -> IO U
cu = coerce

firstFuuzetsu :: Array Int -> IO Int
firstFuuzetsu ar = do
    ref <- newIORefU 0
    U () <- foldMapFuuzetsu (\x -> cu $ modifyIORefU ref (+ x)) ar
    readIORefU ref

firstTreeOwl :: Array Int -> IO Int
firstTreeOwl ar = do
    ref <- newIORefU 0
    U () <- foldMapTreeOwl (\x -> cu $ modifyIORefU ref (+ x)) ar
    readIORefU ref
```

Comment (treeowl) [reply Data/HashMap/Array.hs]: Huh. I'll have to take a look at the Core myself and see if I've done something silly I should correct. If not, we'll just go with yours.

**Comment by sjakobi** [comment]
[A blogpost](https://www.fosskers.ca/blog/container-sizes-en.html) has generated some interest in improving the `Foldable` instances here, specifically `length`.

Does this PR need more work before it can be merged? If so, could the uncontroversial parts be merged and released separately, so users can benefit from them?

**Comment by emilypi** [comment]
Bump on this - @Fuuzetsu does this need more work?

File: Data/HashMap/Array.hs (hunk 1 above)
Comment (treeowl) [reply Data/HashMap/Array.hs:457]: Going with yours for now.

**Pull request merged**